packages/gatsby-core-utils/src/fetch-remote-file.ts

```diff
@@ -1,541 +1,244 @@
 
-  // when a cache entry is present we wait until it completes
-  const result = await new Promise<string | undefined>((resolve, reject) => {
-    pollUntilComplete(cache, url, BUILD_ID, (err, result) => {
-      if (err) {
-        return reject(err)
-      }
+  if (!cache && !directory) {
+    throw new Error(`You must specify either a cache or a directory`)
+  }
 
-      return resolve(result)
-    })
-  })
+  const fetchFileMutex = createMutex(`gatsby-core-utils:fetch:${url}`)
+  await fetchFileMutex.acquire()
 
-  if (result) {
-    return result
-  }
+  // Fetch the file.
+  try {
+    const inFlightValue = getInFlightObject(url, BUILD_ID)
+    if (inFlightValue) {
+      return inFlightValue
+    }
 
-  await cache.set(cacheIdForWorkers(url), {
-    status: `pending`,
-    result: null,
-    workerId: WORKER_ID,
-    buildId: BUILD_ID,
-  })
+    const cachedEntry = await storage.remoteFileInfo.get(url)
 
-  // See if there's response headers for this url
-  // from a previous request.
-  const { headers: cachedHeaders, digest: originalDigest } =
-    (await cache.get(cacheIdForHeaders(url))) ?? {}
-  const headers = { ...httpHeaders }
-  if (cachedHeaders && cachedHeaders.etag) {
-    headers[`If-None-Match`] = cachedHeaders.etag
-  }
+    // See if there's response headers for this url
+    // from a previous request.
+    const headers = { ...httpHeaders }
+    if (cachedEntry?.headers?.etag) {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      headers[`If-None-Match`] = cachedEntry.headers.etag
+    }
 
-  // Add htaccess authentication if passed in. This isn't particularly
-  // extensible. We should define a proper API that we validate.
-  const httpOptions: Options = {}
-  if (auth && (auth.htaccess_pass || auth.htaccess_user)) {
-    httpOptions.username = auth.htaccess_user
-    httpOptions.password = auth.htaccess_pass
-  }
+    // Add htaccess authentication if passed in. This isn't particularly
+    // extensible. We should define a proper API that we validate.
+    const httpOptions: Options = {}
+    if (auth && (auth.htaccess_pass || auth.htaccess_user)) {
+      httpOptions.username = auth.htaccess_user
+      httpOptions.password = auth.htaccess_pass
+    }
 
-  // Create the temp and permanent file names for the url.
-  let digest = createContentDigest(url)
+    if (!name) {
+      name = getRemoteFileName(url)
+    }
 
-  // if worker id is present - we also append the worker id until we have a proper mutex
-  if (IS_WORKER) {
-    digest += `-${WORKER_ID}`
-  }
+    if (!ext) {
+      ext = getRemoteFileExtension(url)
+    }
 
-  if (!name) {
-    name = getRemoteFileName(url)
-  }
-  if (!ext) {
-    ext = getRemoteFileExtension(url)
-  }
+    await fs.ensureDir(fileDirectory as string)
 
-  const tmpFilename = createFilePath(pluginCacheDir, `tmp-${digest}`, ext)
+    const digest = createContentDigest(url)
+    // TODO remove temp file- now with mutex we don't have to
```

Comment:
TODO left here

Reply:
I won't be tackling the TODO in this PR as I'm not sure if it's beneficial

Reply:
Yes, seems like got still doesn't support 304 so I've kept what we had but made it async to have better support on windows